File: recursion/power.py
class Solution:
    def pow(self, x: float, n: int) -> float:
        if n == 0:
            return 1
        if n & 1 == 0:
            p = self.pow(x, n // 2)
            return p * p
        return x * self.pow(x, n - 1)

    def myPow(self, x: float, n: int) -> float:
        if n < 0:
            return (1 / self.pow(x, -n))
        return (self.pow(x, n))

File: recursion/power.c
double _pow(double x, int n)
{
    if (n == 0)
        return 1;
    double half = _pow(x, n / 2);
    if ((n & 1) == 0)
        return half * half;
    else
        return half * half * x;
}

double myPow(double x, int n)
{
    if (n >= 0)
        return (_pow(x, n));
    if (n == INT_MIN)
        return (1 / (x * _pow(x, -(n + 1))));
    return (1 / _pow(x, -n));
}
